xdsl/dialects/mod_arith.py

```diff
@@ -0,0 +1,71 @@
+"""
+mod_arith is a dialect implementing modular arithmetic, originally
+implemented as part of the HEIR project (https://github.com/google/heir/tree/main).
+"""
+
+from abc import ABC
+from typing import Annotated, Generic, TypeVar
+
+from xdsl.dialects.arith import signlessIntegerLike
+from xdsl.dialects.builtin import AnyIntegerAttr
+from xdsl.ir import Attribute, Dialect, Operation, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    pass
+
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class BinaryOp(ModArithOp, ABC, Generic[_T]):
+    """
+    Simple binary operation
+    """
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+    lhs = operand_def(T)
+    rhs = operand_def(T)
+    output = result_def(T)
+
+    assembly_format = "$lhs `,` $rhs attr-dict `:` type($output)"
+    traits = frozenset((Pure(),))
+
+    def __init__(
+        self,
+        lhs: SSAValue | Operation,
+        rhs: SSAValue | Operation,
+        result_type: Attribute | None,
+    ):
+        if result_type is None:
+            result_type = SSAValue.get(lhs).type
+
+        super().__init__(operands=[lhs, rhs], result_types=[result_type])
+
+
+class BinaryModuloOp(BinaryOp[_T], ABC, Generic[_T]):
+    """
+    Adds the modulus attribute
+    """
+
+    modulus = attr_def(AnyIntegerAttr)
```

Comment:
Tablegen "attributes" are actually properties (not sure if you were trying to be compatible.

Reply:
Yeah, upstream has modulus as an attribute not a property

Reply:
In [here](https://github.com/google/heir/blob/main/lib/Dialect/ModArith/IR/ModArithOps.td) I am 90% sure it is a property

Reply:
Seeing the assembly format only mentions `attr-dict`:

> `let assemblyFormat = "operands attr-dict `:` type($output)";`

[ModArithOps.td:75](https://github.com/google/heir/blob/4ca0a98a682a0f2f2569bca50007a3b2d5e50ca6/lib/Dialect/ModArith/IR/ModArithOps.td#L75C1-L75C63)

And the op is printed as `%reduce = mod_arith.reduce %c4 { modulus = 17 } : i10` in [tests/mod_arith/syntax.mlir:37](https://github.com/google/heir/blob/4ca0a98a682a0f2f2569bca50007a3b2d5e50ca6/tests/mod_arith/syntax.mlir#L37)

I'm 99.9% sure it's an attribute.

Reply:
I tried to install the tool for an hour and failed, but the real test will be printing generic syntax

Reply:
It took me all evening and installing a virtual machine but can confirm that it is a property:
![image](https://github.com/user-attachments/assets/a0b0a1d5-322b-435a-bc97-55c7af066cbe)

Reply:
I'm not fully clear on what I need to do here - should I change attr_def to prop_def?